iset.mm

```diff
@@ -63822,36 +63848,46 @@ all reals (greatest real number) for which all positive reals are greater.
     $( Right closure of disjoint union.  (Contributed by Jim Kingdon,
        21-Jun-2022.) $)
     djurcl $p |- ( C e. B -> ( inr ` C ) e. ( A |_| B ) ) $=
-      ( vx wcel cinr cfv c1o cop cdju cvv csn cxp wceq elex com 1onn elexi snid
-      opelxpi mpan cv opeq2 df-inr fvmptg syl2anc c0 cun elun2 df-dju syl6eleqr
-      syl eqeltrd ) CBEZCFGZHCIZABJZUNCKEUPHLZBMZEZUOUPNCBOHUREUNUTHHPQRSHCURBT
-      UAZDCHDUBZIUPKUSFVBCHUCDUDUEUFUNUPUGLAMZUSUHZUQUNUTUPVDEVAUPUSVCUIULABUJU
-      KUM $.
+      ( wcel cinr cfv c1o cop cdju cvv csn cxp wceq elex 1oex snid opelxpi mpan
+      vx cv opeq2 df-inr fvmptg syl2anc cun elun2 syl df-dju syl6eleqr eqeltrd
+      c0 ) CBDZCEFZGCHZABIZULCJDUNGKZBLZDZUMUNMCBNGUPDULURGOPGCUPBQRZSCGSTZHUNJ
+      UQEUTCGUASUBUCUDULUNUKKALZUQUEZUOULURUNVBDUSUNUQVAUFUGABUHUIUJ $.
   $}
 
   ${
-    $d x y $.
-    $( The left injection function on all sets is one to one and onto.
-       (Contributed by Jim Kingdon, 22-Jun-2022.) $)
-    djulf1o $p |- inl : _V -1-1-onto-> ( { (/) } X. _V ) $=
-      ( vx vy cvv c0 csn cxp cinl wf1o wtru cv cop c2nd cfv df-inl wcel opelxpi
-      0ex snid adantl wceq mpan xp2nd c1st 1st2nd2 xp1st elsni syl opeq1d eqtrd
-      wb eqeq2d eqcom eqid vex opth mpbiran 3bitr3g bicomd ad2antll f1o2d trud
-      ) CDEZCFZGHIABCVCDAJZKZBJZLMZGANVDCOZVEVCOZIDVBOVHVIDQRDVDVBCPUASVFVCOZVG
-      COIVFVBCUBSVJVDVGTZVFVETZUJIVHVJVLVKVJVEVFTVEDVGKZTZVLVKVJVFVMVEVJVFVFUCM
-      ZVGKVMVFVBCUDVJVODVGVJVOVBOVODTVFVBCUEVODUFUGUHUIUKVEVFULVNDDTVKDUMDVDDVG
-      QAUNUOUPUQURUSUTVA $.
-
-    $( The right injection function on all sets is one to one and onto.
-       (Contributed by Jim Kingdon, 22-Jun-2022.) $)
-    djurf1o $p |- inr : _V -1-1-onto-> ( { 1o } X. _V ) $=
-      ( vx vy cvv c1o csn cxp cinr wf1o wtru cop c2nd cfv df-inr wcel com snidg
-      cv 1onn adantl wceq ax-mp opelxpi mpan xp2nd c1st 1st2nd2 xp1st elsni syl
-      wb opeq1d eqtrd eqeq2d eqcom eqid elexi vex opth mpbiran 3bitr3g ad2antll
-      bicomd f1o2d trud ) CDEZCFZGHIABCVFDAQZJZBQZKLZGAMVGCNZVHVFNZIDVENZVKVLDO
-      NVMRDOPUADVGVECUBUCSVIVFNZVJCNIVIVECUDSVNVGVJTZVIVHTZUJIVKVNVPVOVNVHVITVH
-      DVJJZTZVPVOVNVIVQVHVNVIVIUELZVJJVQVIVECUFVNVSDVJVNVSVENVSDTVIVECUGVSDUHUI
-      UKULUMVHVIUNVRDDTVODUODVGDVJDORUPAUQURUSUTVBVAVCVD $.
+    $d X x y $.
+    djuf1olem.1 $e |- X e. _V $.
+    djuf1olem.2 $e |- F = ( x e. _V |-> <. X , x >. ) $.
+    $( Lemma for ~ djulf1o and ~ djurf1o .  (Contributed by BJ and Jim Kingdon,
+       4-Jul-2022.) $)
+    djuf1olem $p |- F : _V -1-1-onto-> ( { X } X. _V ) $=
```

Comment:
As stated, the advantages of this are to remove the feeling of proving things twice.

The disadvantage is that reading the proof means following one more layer of indirection.

I guess if we went by the "rule of three" from https://wou.edu/las/cs/csclasses/cs161/Lectures/rulesofthumb.html we'd say this is only two, not three, occurences.

But this sort of thing should be (in my opinion) a bit case by case and in this situation I don't really have a strong opinion one way or the other.

Reply:
I like these lemmas which extract the essentials from the theorems/proofs for inl/inr. Not only they shorten the proofs in sum, and provide the essentials in a more explicit way, they may also be used in other contexts.

Reply:
Thanks for the link: I read with interest all rules. It looks like the rule is "if you write it at least thrice, then write a lemma/subprogram". But the converse is not stated ! After all, sometimes, even a part of a proof or code that is not repeated deserves its own lemma, for reasons other than repetition, like @avekens explained.

So I guess I will keep the lemmas already in Main, which offer a global shortening. However, I will not move djucllem to Main, since it does not offer a global shortening. (I might keep it in my mathbox, and probably delete it in a while).

Reply:
> It looks like the rule is "if you write it at least thrice, then write a lemma/subprogram". But the converse is not stated !

Well https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming) states the converse but to treat these as rigid rules kind of misses the point (in my opinion). As important as "don't remove duplication if there are only two copies" is a sense for why you might not (for example, you might generalize in a less useful way if you don't have enough sense for how it might be used).